_posts/2021-05-04-intro-to-criteria-api.md

```diff
@@ -0,0 +1,473 @@
+- Metamodel API: **It provides the ability to retrieve details on persistent object models, 
+so it's kind of what reflection is towards general java types**
+- Specification: **It's what makes Criteria API predicates reusable and eliminates boilerplates**
+
+Let's define our entity object `Product` first:
+
+````java
+@Entity
+@Table(name = "t_product")
+public class Product {
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.AUTO)
+  private Long id;
+
+  @Column(name = "c_name")
+  private String name;
+
+  @Column(name = "c_category")
+  private String category;
+
+  @Column(name = "c_price")
+  private String price;
+
+  //  getter and setter
+
+}
+````
+
+## Create Query with Criteria Api
+We're going to retrieve products that have specific names and categories using Criteria API.
+
+````java
+@Repository
+public class ProductRepositoryImpl {
+        
+  @PersistenceContext
+  EntityManager em;
+
+  public List<Product> findProductsByNameAndCategory(
+                                            String name,
+                                            String category) {
+      
+  CriteriaBuilder cBuilder = em.getCriteriaBuilder();
+  CriteriaQuery<Product> cQuery = cBuilder.createQuery(Product.class);
+
+  Root<Product> root = cQuery.from(Product.class);
+
+  /* Conditions **/
+  Predicate namePredicate = cBuilder.
+    like(root.get("name"), '%' + (name == null ? "" : name) + '%');
+  Predicate categoryPredicate = cBuilder.
+    equal(root.get("category"), (category == null ? Category.TOOLS : category));
+
+  cQuery.where(namePredicate, categoryPredicate);
+
+  /* Create Query object **/
+  TypedQuery<Product> query = em.createQuery(cQuery);
+
+  return query.getResultList();
+  }
+}
+````
+
+The above code's description is as follows:
+
+* First two lines of the method shows how to get `CriteriaBuilder` and `CriteriaQuery` to create different parts of the query
+* Variable `root` is like `FROM table_name` in our query declaration introduced to the `cQuery` object
+* Variables `namePredicate` and `categoryPredicate` are defined as conditions in the query. Only applied if used
+as parameters of `where` method of `cQuery` object.
+* We finally create the `TypedQuery` object with `em.createQuery(cQuery)`. `TypedQuery` interface is a
+ subclass of `Query` interface.
+ 
+ Objects `namePredicate` and `categoryPredicate` are predicates that construct `where` clause.
+
+## JPA Metamodel
+In the above code example, We used the name of the fields of the entity classes in the predicates like `"name"` and `"category"`.
+**Using string names of entities in the predicates has some flaws**:
+- **It's hard to remember the name so we might need to look it up**
+- **Query will need refactoring if the column's name change later**
+- **It increases the chance of errors in runtime because of the possible mistype** 
+
+**Using JPA Metamodels fixes the above problems**. **It helps us to avoid using the column's name**. It does
```

Comment:
There's too much bold in this article in general. Please go over the text and think about which sentences should really stand out and be bold. These sentences should stand on their own and not need the context of the other sentences (they maybe need some editing, as a sentence starting with "It helps..." does not have enough context on its own - instead we could say "The JPA Metamodel helps ...".